File: sdks/typescript/src/protocol.ts
// Wire-protocol types for the Croniq runner API.
//
// Field names are snake_case to match the JSON shape in `openapi.yaml`.
// These types are sent and received verbatim over HTTP — no field
// transformation happens at the client boundary, so consumer code that
// touches `metadata` will see the original snake_case keys the server sent.

export interface PollRequest {
  runner_id: string;
  capabilities: string[];
  max_inflight: number;
  inflight: string[];
  instance_id?: string;
  tags: string[];
}

export interface WorkAssignment {
  execution_id: string;
  job_key: string;
  fire_at: string;
  /**
   * Original logical fire time (RFC 3339). Absent when the server predates
   * the field — consumers must not fall back to `fire_at`.
   */
  scheduled_for?: string;
  attempt: number;
  metadata: unknown;
  timeout?: string;
}

export interface PollResponse {
  work: WorkAssignment[];
  cancel: string[];
}

export type AckStatus = 'success' | 'failure';

export interface AckRequest {
  runner_id: string;
  execution_id: string;
  status: AckStatus;
  error?: string;
  duration_ms?: number;
  attempt: number;
}

export interface RenewRequest {
  runner_id: string;
  execution_id: string;
}

export interface RegisterJobRequest {
  job_key: string;
  schedule: string;
  timezone?: string;
  timeout?: string;
  runner_id?: string;
  capabilities: string[];
  description?: string;
}

export interface RegisterJobResponse {
  job_key: string;
  trigger_id?: string;
  status?: 'registered' | 'skipped_dsl_precedence';
}

export type WorkEventLevel = 'trace' | 'debug' | 'info' | 'warn' | 'error';

export interface WorkEvent {
  level?: WorkEventLevel;
  message: string;
  fields?: Record<string, string>;
}

/**
 * Wire request for `POST /v1/trigger` (producer side). Optional fields are
 * omitted from the JSON body when unset — a producer never emits a key the
 * caller didn't supply (`JSON.stringify` drops `undefined` values).
 */
export interface TriggerRequest {
  job_key: string;
  metadata?: Record<string, unknown>;
  require?: string[];
  prefer?: string[];
  timeout?: string;
  idempotency_key?: string;
}

/**
 * Wire response of `POST /v1/trigger`. `deduplicated` is sent by servers that
 * support trigger idempotency keys (#279); older servers omit it and the
 * client defaults it to `false`.
 */
export interface TriggerResponse {
  execution_id: string;
  queued: number;
  deduplicated?: boolean;
}

File: sdks/python/src/croniq_runner/_protocol.py
"""Pydantic DTOs for the Croniq runner wire protocol.

Field names mirror `openapi.yaml` exactly: snake_case on the wire, snake_case
in Python. Optional fields use ``= None`` defaults; ``model_dump(by_alias=True,
exclude_none=True)`` produces the JSON shape the server accepts.
"""

from __future__ import annotations

from typing import Any

from pydantic import BaseModel, ConfigDict, Field


class _Model(BaseModel):
    model_config = ConfigDict(
        extra="ignore",
        populate_by_name=True,
        str_strip_whitespace=False,
    )


class PollRequest(_Model):
    """POST /v1/work/poll request body."""

    runner_id: str
    capabilities: list[str] = Field(default_factory=list)
    max_inflight: int = 1
    inflight: list[str] = Field(default_factory=list)
    instance_id: str | None = None
    tags: list[str] = Field(default_factory=list)


class WorkAssignment(_Model):
    """One unit of work returned by /v1/work/poll."""

    execution_id: str
    job_key: str
    fire_at: str
    # Original logical fire time (RFC 3339). None when the server predates the
    # field — the SDK must not fall back to fire_at.
    scheduled_for: str | None = None
    attempt: int
    metadata: dict[str, Any] = Field(default_factory=dict)
    timeout: str


class PollResponse(_Model):
    """POST /v1/work/poll response body."""

    work: list[WorkAssignment] = Field(default_factory=list)
    cancel: list[str] = Field(default_factory=list)


class AckRequest(_Model):
    """POST /v1/work/ack request body."""

    runner_id: str
    execution_id: str
    status: str  # "success" | "failure"
    error: str | None = None
    duration_ms: int | None = None
    attempt: int


class RenewRequest(_Model):
    """POST /v1/work/renew request body."""

    runner_id: str
    execution_id: str


class WorkEvent(_Model):
    """A structured log event pushed to /v1/work/{id}/events.

    The runner auto-enriches ``fields`` with ``job_key``, ``runner_id`` and
    (when set) ``runner_tags`` before the POST — caller-supplied keys win.
    """

    level: str | None = None
    message: str
    fields: dict[str, str] | None = None


class RegisterJobRequest(_Model):
    """POST /v1/jobs/register request body."""

    job_key: str
    schedule: str
    timezone: str | None = None
    timeout: str | None = None
    runner_id: str | None = None
    capabilities: list[str] = Field(default_factory=list)
    description: str | None = None


class RegisterJobResponse(_Model):
    """POST /v1/jobs/register response body."""

    job_key: str | None = None
    trigger_id: str | None = None
    status: str | None = None  # "registered" | "skipped_dsl_precedence"


class TriggerRequest(_Model):
    """POST /v1/trigger request body (producer side).

    Every field except ``job_key`` is optional; dumping with
    ``exclude_none=True`` omits the ones the caller left unset (the server
    treats an omitted optional and a ``null`` differently — see #283).
    ``metadata`` is arbitrary caller JSON forwarded to the handler verbatim,
    so its values are ``Any``, not strings.
    """

    job_key: str
    metadata: dict[str, Any] | None = None
    require: list[str] | None = None
    prefer: list[str] | None = None
    timeout: str | None = None
    idempotency_key: str | None = None


class TriggerResponse(_Model):
    """POST /v1/trigger response body.

    ``deduplicated`` is sent by servers with trigger-idempotency support
    (#279); older servers omit it and the field defaults to ``False``.
    """

    execution_id: str
    queued: int
    deduplicated: bool = False
